### PR TITLE
enhancement: add ignition dual spec 2/3 support

### PR DESCRIPTION
**Pull request by yuqi-zhang**
Add an enhancement proposal for supporting both ignition spec 2
and 3 for OS provisioning/updating.

Signed-off-by: Yu Qi Zhang <jerzhang@redhat.com>

### CODE REVIEW

File: enhancements/ignition-spec-dual-support.md [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+ - The user is able to recover the cluster and finish the update if the untranslatable configs are manually translated or remove, or roll back to the old version
+ - The user should have received notification that the update will be changing spec version, as well as received necessary documentation on how to recover a failed update
+ - CI tests are put in place to make sure the existing versions can be updated to the new payload
+
+** As a user of openshift, I’d like to install from a spec 2 bootimage andimmediately update to a spec 3 payload **
+
+Acceptance criteria:
+ - Essentially the same as story 1
+
+** As a user of Openshift, I’d like to install a fresh ignition spec 3 cluster **
```
Comment (abhinavdahiya) [inline enhancements/ignition-spec-dual-support.md]: the users just want to install a cluster, imo there is less value in customers asking for spec2 vs spec3

above we mention that the installer can switch to spec3 completely, do we have some idea how we can make sure UPI customers are using the correct boot-images when they get the spec3 ignition configs but maybe try to use a spec3 supporting bootimage

File: enhancements/ignition-spec-dual-support.md [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+** MCO details: **
+
+A spec translator will first be implemented in the MCO, with the ability to
+detect untranslatable configs. The MCO then should be updated to have support
+for both ignition V0S2 and V2S3. Since the MCO is responsible for the current
+cluster nodes, it will be the only place at which spec translation is done.
+The translation will happen when the version of MCO with dual support and
+translator is first deployed; it will detect the existing config being spec 2,
+generate a new renderedconfig based on a translator from spec2 to spec 3, and
+do a “dummy reboot” (if necessary) to bring the cluster to the desired config,
```
Comment (sgreene570) [inline enhancements/ignition-spec-dual-support.md]: What exactly do you mean by "dummy reboot" in this scenario? When would the "dummy reboot" be necessary?

File: enhancements/ignition-spec-dual-support.md [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+ - The workflow remains the same for an IPI cluster
+ - The workflow remains the same for a UPI cluster, minus custom specification changes
+ - The user should have good documentation, based on version, of how to set up user defined configs during install time
+
+** As an admin of an existing spec 3 cluster, I’d like to apply a new machineconfig **
+
+Acceptance criteria:
+ - The machineconfig is applied successfully, if the user has defined a correct spec 3 ignition snippet
+ - The user is properly alerted if they attempt to apply a spec 2 config, and the machineconfig fails to apply
+ - The user is given necessary docs to remove the undesired spec 2 config and to translate it to a spec 3 config
```
Comment (sgreene570) [inline enhancements/ignition-spec-dual-support.md]: What is "the undesired spec 2 config" in this case? The last bullet says any spec 2 config would fail to apply.

File: enhancements/ignition-spec-dual-support.md [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+Extensive testing of all possible paths, especially those outlined in the
+user stories, is critical to the success of this major update. The existing
+CI infrastructure is a good start for upgrade testing. There should be
+additional tests added, especially in the MCO repo, for edge cases as
+described in the user stories, to ensure we never break this behaviour.
+Many existing tests will also have to be updated given the spec change.
+
+
+### Upgrade / Downgrade Strategy
```
Comment (sgreene570) [inline enhancements/ignition-spec-dual-support.md]: Downgrades from spec 3 to spec 2 are off the table, right?

File: enhancements/ignition-spec-dual-support.md (hunk 1 above)
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: I agree, the workflow should look the same for IPI

> do we have some idea how we can make sure UPI customers are using the correct boot-images when they get the spec3 ignition configs but maybe try to use a spec3 supporting bootimage

We'll definitely have to communicate in the docs which preview version has spec 3 support. The installer maybe should add capability to detect if the user is still on an old version and reject the install.

File: enhancements/ignition-spec-dual-support.md (hunk 2 above)
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: Since the cluster machineconfig is saved as a "desiredConfig" hash, I'm not sure the best way to translate that yet. This just means that after the specs are translated, if we need, we can set a new desiredConfig that specifies a successful translation to spec 3, which the cluster needs to reboot to bring the currentConfig up to.

File: enhancements/ignition-spec-dual-support.md (hunk 3 above)
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: Sorry, it just means any spec 2 config in the system. Basically steps to `oc delete machineconfig/xxx` and then apply a new one

File: enhancements/ignition-spec-dual-support.md (hunk 1 above)
Comment (abhinavdahiya) [reply enhancements/ignition-spec-dual-support.md]: > The installer maybe should add capability to detect if the user is still on an old version and reject the install.

at what stage do you think the installer will be able to do that?

in UPI, the installer only outputs ignition files and rest is user controlled process.

File: enhancements/ignition-spec-dual-support.md (hunk 4 above)
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: Yes. We should never have to do that. We do however probably need rollback functionality to recover from bad machineconfig states.

File: enhancements/ignition-spec-dual-support.md (hunk 1 above)
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: Oh true, I guess in this case the best we can do is to alert the user that this build will output spec 3 ignition configs, which is only supported by RHCOS versions x and later?

This goes back to https://github.com/openshift/installer/issues/1399 which might be a better UPI workflow. Alternatively, for example if we decide to do this in 4.5, all 4.5 images + installer generated configs are spec 3 by default only. In that case a user would only run in to a version mistmatch if they try to install a 4.5 cluster with 4.4 bootimages, which they shouldn't do anyways.

Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: Open to suggestions of course

Comment (abhinavdahiya) [reply enhancements/ignition-spec-dual-support.md]: > This goes back to openshift/installer#1399 which might be a better UPI workflow. Alternatively, for example if we decide to do this in 4.5, all 4.5 images + installer generated configs are spec 3 by default only. In that case a user would only run in to a version mistmatch if they try to install a 4.5 cluster with 4.4 bootimages, which they shouldn't do anyways.

> if they try to install a 4.5 cluster with 4.4 bootimages, which they shouldn't do anyways.

that means that we don't have N-1 compatibility, which is not we tell our users! the n-1 should be fine as long as you don't depend on a feature that is in N.

> This goes back to openshift/installer#1399 which might be a better UPI workflow.

I have specifically pushed back on that idea because, that allows us to demand this 4.4 works with 4.4 bootimage only, when that shouldn't be the case for most users. and also something we need to support in our clusters because 4.1 created clusters have new nodes join in at 4.1 and then pivot to say 4.4.

> Open to suggestions of course since UPI in general will be harder to coordinate

+1
as long as we explicitly mention what we are doing for this case, i'm fine.

File: enhancements/ignition-spec-dual-support.md [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+** As the admin of an existing 4.x cluster on spec 2, I’d like to upgrade to the newest version and use ignition spec 3 **
+
+Acceptance criteria:
+ - The update completes without user intervention, if all machine configs existing on the cluster can be directly translated to spec 3
+ - The user receives an alert, if the update is unable to complete due to untranslatable configs
+ - The user is able to recover the cluster and finish the update if the untranslatable configs are manually translated or remove, or roll back to the old version
+ - The user should have received notification that the update will be changing spec version, as well as received necessary documentation on how to recover a failed update
+ - CI tests are put in place to make sure the existing versions can be updated to the new payload
+
+** As a user of openshift, I’d like to install from a spec 2 bootimage andimmediately update to a spec 3 payload **
```
Comment (wking) [inline enhancements/ignition-spec-dual-support.md]: nit: "andimmediately" -> "and immediately"

File: enhancements/ignition-spec-dual-support.md [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+other similar cases, e.g. for auto-deployed metal clusters, the served ignition
+configs must all be updated after a certain point of the bootimage to be able
+to bring up new machines. To mitigate we should communicate this change well in
+advance, and provide methods to translate ignition configs. Failed
+installation/alerting systems must clearly communicate the source of error in
+this case, as well as how to mitigate.
+
+** How to properly detect which ignition version to serve new machines **
+
+This will likely be somewhat difficult to handle, as the wrong version on a
```
Comment (wking) [inline enhancements/ignition-spec-dual-support.md]: Can we teach Ignition to request a version using `Accept` headers?  Looks like it may [already do that][1]?

[1]: https://github.com/coreos/ignition/blob/15b5b04ef0c048eebd1cfe5b6192d7a6fc19d625/internal/resource/url.go#L57

Comment (wking) [reply enhancements/ignition-spec-dual-support.md]: And with the machine-config server switching on `application/vnd.coreos.ignition+json;version=...` and an ability to translate spec-2 configs to spec-3 and spec-3 configs to spec-2, we'd be able to run a full 4.4.z (or whatever z stream) supporting both configs, and warning about 4.5 (or whatever) incompatibility if we saw any spec-2 requests come in.

**Comment by yuqi-zhang** [comment]
Updated per discussions with @crawford, PTAL.

File: enhancements/ignition-spec-dual-support.md [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+translation is completely successful, the MCD will be instructed to use
+the spec 3 configs. Otherwise the MCD will continue using the existing
+spec 2 configs, and apply new spec 2/spec 3 configs. If at any point the
+MCC notices it can fully translate to spec 3, that will be the version we
+use.
+
+The MCS will host both spec 3 and spec 2 configs, with a functionality to
+translate spec 3 down to spec 2 configs, and spec 2 up to spec 3 configs.
+The MCS will first check which ignition spec version a new node supports,
+before serving a config.
```
Comment (runcom) [inline enhancements/ignition-spec-dual-support.md]: is there already something in place that we can leverage? or do we need to check with MAO (or something else?) to provide us with something like that

File: enhancements/ignition-spec-dual-support.md [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+
+ - Support FCoS/OKD directly through this change
+ - API support for MCO, namely switching to RawExt formatted machineconfig
+ objects instead of explicitly referencing ignition, is not considered as
+ part of this proposal
+
+
+## Proposal
+
+This change is multi-component:
```
Comment (runcom) [inline enhancements/ignition-spec-dual-support.md]: @ericavonb brought up that we might need to sync/communicate with KNI as well (iirc)

**Comment by cgwalters** [comment]
> The MCS will first check which ignition spec version a new node supports, before serving a config

See https://github.com/coreos/ignition/pull/889
Which was sketched out initially in https://github.com/openshift/machine-config-operator/issues/492

File: enhancements/ignition-spec-dual-support.md [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+    - [ ] a translator in the Machine Config Controller that can convert spec 2 cluster configuration to spec 3
+    - [ ] a detector in the Machine Config Server that can serve the correct spec version based on bootimage version
+- [ ] The Openshift installer is updated to generate ignition spec 3 configs
+- [ ] Successfully create new Spec 3 only IPI/UPI clusters
+- [ ] Create a migration method for non-compatible (non-translatable configs) Spec 2 clusters to Spec 3
+- [ ] An alerting mechanism is put in place for outdated and incompatible/non-translatable configs
+- [ ] Docs are updated to reflect the new config version
+
+#### Phase 2:
+- [ ] RHCOS bootimage is updated to accept ignition spec 3 configs
```
Comment (runcom) [inline enhancements/ignition-spec-dual-support.md]: spec v3 _only_

**Comment by yuqi-zhang** [comment]
Updated to reflect the new phase 1/2 plans. Specifically, our main objective in phase 1 becomes adding the ability to apply spec 3 machineconfigs. Does not involve installer/RHCOS.

Also note the new `Managing stub master/worker ignition configs` section, which will also be a separate enhancement PR.

File: enhancements/ignition-spec-dual-support.md [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+
+#### Phase 1:
+- [ ] A config translator is created to translate from ignition spec 2 to spec 3
+- [ ] A config translator is created to translate from ignition spec 3 to spec 2
+- [ ] The MCO is updated to support both ignition spec 2 and 3, with:
+    - [ ] MCC rendering spec 3 configs to spec 2
+    - [ ] MCD gaining the ability to process both spec 2 and spec 3 configs
+- [ ] Create tests to apply spec 3 configs to spec 2 clusters successfully
+- [ ] An alerting mechanism is put in place for outdated and incompatible/non-translatable configs
+- [ ] Docs are updated to reflect the new config version
```
Comment (runcom) [inline enhancements/ignition-spec-dual-support.md]: @crawford can you ack this plan if it does look good to you? Not sure who else to bring in

File: enhancements/ignition-spec-dual-support.md [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+machine is on spec 2 (will happen only once for all existing and new nodes,
+when the MCO with dual support is first deployed onto the cluster). This will
+only be required as part of the MCO.
+
+Note that there exists three types of spec 2 configs:
+ - Those that are directly translatable to spec 3. This is the case for all existing IPI configs.
+ - Those that we’re not 100% sure we can directly translate, but we can infer what the user is doing and do a translation 
+ - Those that we’re 100% sure we CANNOT translate directly, and requires user input for us to correctly translate
+
+During phase 1 we should attempt to translate on a best-effort basis. If the
```
Comment (crawford) [inline enhancements/ignition-spec-dual-support.md]: We should clarify that we will only perform the translation of each config if we are certain that it will be successful. “Best-effort” here refers to the act of visiting all configs, but not the translation for each config.

File: enhancements/ignition-spec-dual-support.md [hunk 12]
```diff
@@ -0,0 +1,447 @@
...
+
+To handle both spec versions, as we need the ability to upgrade existing
+clusters, we will create a translator between spec 2 and spec 3. This ensures
+that a cluster only has one “desiredConfig” which will be translated to spec 3
+when the MCO with dual support detects that the existing configuration of a
+machine is on spec 2 (will happen only once for all existing and new nodes,
+when the MCO with dual support is first deployed onto the cluster). This will
+only be required as part of the MCO.
+
+Note that there exists three types of spec 2 configs:
```
Comment (crawford) [inline enhancements/ignition-spec-dual-support.md]: For the purpose of this document, there are two types: those that we are certain we can translate and everything else. We aren’t going to handle the uncertain and certain-failure cases differently (are we?).

File: enhancements/ignition-spec-dual-support.md [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+ - Those that we’re not 100% sure we can directly translate, but we can infer what the user is doing and do a translation 
+ - Those that we’re 100% sure we CANNOT translate directly, and requires user input for us to correctly translate
+
+During phase 1 we should attempt to translate on a best-effort basis. If the
+cluster can be directly translated to spec 3, we will do so and use the now
+v3 spec version config in the MCD. If not, we will support both, and alert
+the user that there are untranslatable configs.
+
+During phase two, we should fail updates unless the cluster is fully on spec
+3 config. This effectively means that UPI clusters are at risk when upgrading
```
Comment (crawford) [inline enhancements/ignition-spec-dual-support.md]: Why are UPI clusters at risk?

File: enhancements/ignition-spec-dual-support.md [hunk 14]
```diff
@@ -0,0 +1,447 @@
...
+
+Phase 1:
+
+No changes needed.
+
+Phase 2:
+
+The bootstrap ignition configs are updated to be spec 3. The stub ignition
+configs for master/worker nodes are updated to spec 3, and referece a different
+endpoint of the MCS which will serve ignition spec 3 configs. RHCOS images pinned
```
Comment (crawford) [inline enhancements/ignition-spec-dual-support.md]: The endpoint should be the same. Ignition already uses an Accept header to specify which spec versions it can process.

File: enhancements/ignition-spec-dual-support.md [hunk 15]
```diff
@@ -0,0 +1,447 @@
...
+Phase 2:
+
+All machineconfigs are switched to spec 3. Spec 2 support will exist in the form
+of served ignition configs to new nodes joining the cluster that have ignition
+v0 in the bootimage. MCS will handle this with 2 endpoints.
+
+The MCO will flat out reject spec 2 configs, and refuse to upgrade clusters
+that have spec 2 bits.
+
+The MCO will also throw an alert upon: seeing a spec 2 machineconfig applied to
```
Comment (crawford) [inline enhancements/ignition-spec-dual-support.md]: I don’t think we want to bother with alerts here. We can just validate the MachineConfig before the CR is committed.

File: enhancements/ignition-spec-dual-support.md [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+installer and underlying control plane/bootstrap configuration, as well as RHEL 
+CoreOS (Henceforth RHCOS) package version will also be updated.
+
+This overall migration will take part as a two-phase process:
+
+Phase 1/OCP 4.4:
+The objective of this phase is to allow users to apply machineconfigs with
+ignition spec 3 support to new and existing clusters. The breakdown is:
+
+ - a tool is created to translate between ignition spec versions
```
[the remark this answers is not in the document]
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: MCO makes use directly of ignition typing via vendoring (currently 2.2), so we'd need a method to translate the typing on the system from 2->3, unless you think the MCO should be on 2 indefinitely for some reason

File: enhancements/ignition-spec-dual-support.md [hunk 17]
```diff
@@ -0,0 +1,447 @@
...
+
+
+Phase 2/OCP 4.X:
+All new installs will be on spec 3 only. Existing clusters that have non spec 3
+ignition config machineconfigs will not be allowed to update to the new version.
+
+ - RHCOS bootimages switches to only accept ignition spec 3 configs
+ - The Openshift installer is updated to generate spec 3 configs
+ - Remaining MC* components generate spec 3 only
+ - MCO enforces that all configs are spec 3 before allowing the CVO to start the update
```
[the remark this answers is not in the document]
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: This has not been fleshed out yet. I recall discussion regarding using unupgradable flag for this

File: enhancements/ignition-spec-dual-support.md [hunk 18]
```diff
@@ -0,0 +1,447 @@
...
+
+
+### Goals
+
+#### Phase 1:
+- [ ] A config translator is created to translate from ignition spec 2 to spec 3
+- [ ] A config translator is created to translate from ignition spec 3 to spec 2
+- [ ] The MCO is updated to support both ignition spec 2 and 3, with:
+    - [ ] MCC rendering spec 3 configs to spec 2
+    - [ ] MCD gaining the ability to process both spec 2 and spec 3 configs
```
[the remark this answers is not in the document]
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: I guess not, perhaps I phrased this poorly, will revise

File: enhancements/ignition-spec-dual-support.md [hunk 19]
```diff
@@ -0,0 +1,447 @@
...
+- [ ] An alerting mechanism is put in place for outdated and incompatible/non-translatable configs
+- [ ] Docs are updated to reflect the new config version
+
+#### Phase 1 (optional):
+- [ ] MCC attempts to translate all machineconfigs with v2 ignition config to v3, leaving those which cannot be converted and warning the user of future changes
+- [ ] MCC learns to generate v3 initial master/worker ignition configs
+
+#### Phase 2:
+- [ ] The MCO gains the ability to manage installer-generated stub master/worker ignition configs (separate enhancement proposal)
+- [ ] ignition-dracut spec 2 and spec 3 diffs are aligned
```
[the remark this answers is not in the document]
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: There's a non-trivial delta between https://github.com/coreos/ignition-dracut/ and https://github.com/coreos/ignition-dracut/tree/spec2x/ , the latter of which is being used for RHCOS today

File: enhancements/ignition-spec-dual-support.md (hunk 13 above)
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: More likely the user supplied their own ignition config customizations

**Comment by yuqi-zhang** [comment]
Added some fixes.

>Why aren’t we using spec 3 on new clusters in phase 1?

This will require non-trivial changes to the installer and MCO which the team doesn't think we have enough time for in the 4.4 timeframe

File: enhancements/ignition-spec-dual-support.md (hunk 12 above)
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: Updated

File: enhancements/ignition-spec-dual-support.md [hunk 20]
```diff
@@ -0,0 +1,447 @@
...
+
+### Goals
+
+#### Phase 1:
+- [ ] A config translator is created to translate from ignition spec 2 to spec 3
+- [ ] A config translator is created to translate from ignition spec 3 to spec 2
+- [ ] The MCO is updated to support both ignition spec 2 and 3, with:
+    - [ ] MCC rendering spec 3 configs to spec 2
+    - [ ] MCD gaining the ability to process both spec 2 and spec 3 configs
+- [ ] Create tests to apply spec 3 configs to spec 2 clusters successfully
```
[the remark this answers is not in the document]
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: That could maybe work as well, testing has also not been fleshed out

File: enhancements/ignition-spec-dual-support.md [hunk 21]
```diff
@@ -0,0 +1,447 @@
...
+Ignition released v2.0.0 (stable) on Jun 3rd, 2019, which has an updated
+specification format (ignition spec version 3, henceforth “Spec 3”). This
+change includes some important reworks for RHCOS provisioning, most
+importantly the ability to specify and mount other filesystems and fixing
+issues where ignition v0 spec was not declarative. Particularly, this is
+required to support having /var on a separate partition or disk, an important
+requirement for security/compliance purposes in OCP. The existing version on
+RHCOS systems (ignition version v0.33) carries a spec version (spec version 2,
+henceforth “Spec 2”) that is not compatible with Spec 3. Thus we would like to
+update the ignition version on RHCOS/Installer/MCO to make use of the changes.
```
[the remark this answers is not in the document]
Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: If only ignition (the package) is updated to handle both, all MachineConfigs applied after firstboot will not be affected by the change. The MCO does e.g. file writing itself based on machineconfigs, which use ignition typing.

**Comment by runcom** [comment]
ping @openshift/openshift-architects

it does lgtm, can any of you approve?

/lgtm

**Comment by runcom** [comment]
damn, I thought I couldn't approve but I still think this is gonna be the way we all agreed on - @crawford since you reviewed this, please take a last look at this.

File: enhancements/ignition-spec-dual-support.md [hunk 22]
```diff
@@ -0,0 +1,448 @@
...
+spec 2 configs, and apply new spec 2/spec 3 configs.
+
+The MCS served configs generated by the MCC will be downtranslated by the
+MCC to spec 2. The MCS is not updated to serve spec 3 configs yet.
+
+Phase 2:
+
+All machineconfigs are switched to spec 3. Spec 2 support will exist in the form
+of served Ignition configs to new nodes joining the cluster that have Ignition
+v0 in the bootimage. MCS will handle this with 2 endpoints.
```
Comment (crawford) [inline enhancements/ignition-spec-dual-support.md]: > MCS will handle this with 2 endpoints.

This should be a single endpoint. Ignition already uses the Accept HTTP header to specify which configs it can accept and the order of preference.

Comment (runcom) [reply enhancements/ignition-spec-dual-support.md]: cc @yuqi-zhang this is related also to the ignition endpoint contained in that secret created by the installer (and unmanaged)

Comment (yuqi-zhang) [reply enhancements/ignition-spec-dual-support.md]: I can update the proposal. We would still first need an option to modify the secret anyways.
